assert/assertion_order_test.go

```diff
@@ -116,3 +116,84 @@ func TestLessOrEqual(t *testing.T) {
 		t.Error("Greater should return false")
 	}
 }
+
+func Test_compareTwoValuesDifferentValuesTypes(t *testing.T) {
+	mockT := new(testing.T)
+
+	for _, currCase := range []struct {
+		v1            interface{}
+		v2            interface{}
+		compareResult bool
+	}{
+		{v1: 123, v2: "abc"},
+		{v1: "abc", v2: 123456},
+		{v1: float64(12), v2: "123"},
+		{v1: "float(12)", v2: float64(1)},
+	} {
+		compareResult := compareTwoValues(mockT, currCase.v1, currCase.v2, []CompareType{compareLess, compareEqual, compareGreater}, "testFailMessage")
+		if compareResult {
+			t.Errorf("Values %s and %s should be different kinds", currCase.v1, currCase.v2)
+		}
+	}
+}
+
+func Test_compareTwoValuesNotComparableValues(t *testing.T) {
+	mockT := new(testing.T)
+
+	type CompareStruct struct {
+	}
+
+	for _, currCase := range []struct {
+		v1 interface{}
+		v2 interface{}
+	}{
+		{v1: CompareStruct{}, v2: CompareStruct{}},
+		{v1: map[string]int{}, v2: map[string]int{}},
+		{v1: make([]int, 5, 5), v2: make([]int, 5, 5)},
+	} {
+		compareResult := compareTwoValues(mockT, currCase.v1, currCase.v2, []CompareType{compareLess, compareEqual, compareGreater}, "testFailMessage")
+		if compareResult {
+			t.Errorf("Values %s and %s should be not comparable", currCase.v1, currCase.v2)
+		}
+	}
+}
+
+func Test_compareTwoValuesCrrectCompareResult(t *testing.T) {
+	mockT := new(testing.T)
+
+	for _, currCase := range []struct {
+		v1           interface{}
+		v2           interface{}
+		compareTypes []CompareType
+	}{
+		{v1: 1, v2: 2, compareTypes: []CompareType{compareLess}},
+		{v1: 1, v2: 2, compareTypes: []CompareType{compareLess, compareEqual}},
+		{v1: 2, v2: 2, compareTypes: []CompareType{compareGreater, compareEqual}},
+		{v1: 2, v2: 2, compareTypes: []CompareType{compareEqual}},
+		{v1: 2, v2: 1, compareTypes: []CompareType{compareEqual, compareGreater}},
+		{v1: 2, v2: 1, compareTypes: []CompareType{compareGreater}},
+	} {
+		compareResult := compareTwoValues(mockT, currCase.v1, currCase.v2, currCase.compareTypes, "testFailMessage")
+		if !compareResult {
+			t.Errorf("Values %d and %d is not compared correctly", currCase.v1, currCase.v2)
+		}
+	}
+}
+
+func Test_containsValue(t *testing.T) {
+	for _, currCase := range []struct {
+		values []CompareType
+		value  CompareType
+		result bool
+	}{
+		{values: []CompareType{compareGreater}, value: compareGreater, result: true},
+		{values: []CompareType{compareGreater, compareLess}, value: compareGreater, result: true},
+		{values: []CompareType{compareGreater, compareLess}, value: compareLess, result: true},
+		{values: []CompareType{compareGreater, compareLess}, value: compareEqual, result: false},
+	} {
+		compareResult := containsValue(currCase.values, currCase.value)
+		if compareResult != currCase.result {
```

Comment:
We can dogfood our own `assert.Equal` function here :)

Reply:
@boyan-soubachov  yes, it's tricky. Do you think is it OK to use our own `assert.Equal` ? which is potentially should be tested inside `testify`?

Reply:
Oh yes, we actively encourage it :)

2nd paragraph of https://github.com/stretchr/testify/blob/master/README.md#contributing

`Equal` is well tested and if we use it ourselves, we're more likely to find any issues with it _before_ any of our external users do. This means we can also fix it before it becomes a common problem for our external users.

Reply:
@boyan-soubachov Yes, I see.. Fixed :)